pkg/controller/v1alpha1/autoscaler/autoscaler_controller.go

```diff
@@ -0,0 +1,215 @@
+		APIVersion:         scaler.APIVersion,
+		Kind:               scaler.Kind,
+		UID:                scaler.GetUID(),
+		Name:               scaler.Name,
+		Controller:         pointer.BoolPtr(true),
+		BlockOwnerDeletion: pointer.BoolPtr(true),
+	}
+
+	// Fetch the child resources list from the corresponding workload
+	resources, err := util.FetchWorkloadChildResources(r.ctx, log, r, workload)
+	if err != nil {
+		log.Error(err, "Error while fetching the workload child resources", "workload", workload.UnstructuredContent())
+		r.record.Event(eventObj, event.Warning(util.ErrFetchChildResources, err))
+		return util.ReconcileWaitResult, util.PatchCondition(r.ctx, r, &scaler,
+			cpv1alpha1.ReconcileError(fmt.Errorf(util.ErrFetchChildResources)))
+	}
+	resources = append(resources, workload)
+
+	targetWorkloadSetFlag := false
+	for _, res := range resources {
+		resPatch := client.MergeFrom(res.DeepCopyObject())
+		refs := res.GetOwnerReferences()
+		for i, r := range refs {
+			if *r.Controller {
+				refs[i].Controller = pointer.BoolPtr(false)
+			}
+		}
+		refs = append(refs, ownerReference)
+		res.SetOwnerReferences(refs)
+		if err := r.Patch(r.ctx, res, resPatch, client.FieldOwner(scaler.GetUID())); err != nil {
+			log.Error(err, "Failed to set ownerReference for child resource")
+			return util.ReconcileWaitResult,
+				util.PatchCondition(r.ctx, r, &scaler, cpv1alpha1.ReconcileError(
+					errors.Wrap(err, "Failed to set ownerReference for child resource")))
+		}
+		if !targetWorkloadSetFlag && (res.GetKind() == "Deployment" || res.GetKind() == "StatefulSet") {
+			scaler.Spec.TargetWorkload = v1alpha1.TargetWorkload{
+				APIVersion: res.GetAPIVersion(),
+				Kind:       res.GetKind(),
+				Name:       res.GetName(),
+			}
+			targetWorkloadSetFlag = true
+		}
+	}
+	// if there is no child resource or no child resource kind is deployment or statefuset, set the workload as target workload
+	if len(resources) == 0 && !targetWorkloadSetFlag {
+		scaler.Spec.TargetWorkload = v1alpha1.TargetWorkload{
+			APIVersion: workload.GetAPIVersion(),
+			Kind:       workload.GetKind(),
+			Name:       workload.GetName(),
+		}
+	}
+
+	namespace := req.NamespacedName.Namespace
+
+	// TODO(zzxwill) compare two scalers and adjust the target replicas
+
+	if err := r.scaleByHPA(scaler, namespace, log); err != nil {
+		return ReconcileWaitResult, err
+	}
+
+	if err := r.scaleByKEDA(scaler, namespace, log); err != nil {
+		return ReconcileWaitResult, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *AutoscalerReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	if err := r.buildConfig(); err != nil {
+		return err
+	}
+	r.ctx = context.Background()
+	r.record = event.NewAPIRecorder(mgr.GetEventRecorderFor("Autoscaler")).
+		WithAnnotations("controller", "Autoscaler")
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1alpha1.Autoscaler{}).
+		Complete(r)
+}
+
+func (r *AutoscalerReconciler) buildConfig() error {
```

Comment:
Why do we need this in the controller? The config should be passed in through the constructer